### PR TITLE
First patcher for Gutenberg blocks content after conversion

### PR DESCRIPTION
**Pull request by iuravic**
Created the first converted blocks' content patcher, `ImgPatcher`, which patches the lost `height` attribute, and solves the first part of the ["Content Loss #1" case](https://newsprojectp2.wordpress.com/2019/07/17/newspack-content-migration-spot-check-current-status/#comment-653). It serves to lay down the foundation of creating new patches (adding new patcher classes to the `lib/content-patcher/patchers/` folder, just like the `class-img-patcher.php` here).

The service structure consists of two main entities:
 - the `\NewspackContentConverter\ContentPatcher\PatchHandler` -- a registry of all available patchers. It executes them one by one. _(It might actually be more properly named `PatchRegistry` instead of `Handler`.)_ It is currently used in the `patch_converted_blocks_content()` in the Plugin's main class `lib/class-converter.php`,
 - the `\NewspackContentConverter\ContentPatcher\Patchers\ImgPatcher` -- patches the lost `height` attribute of the `` element. By its structure, it extends the `NewspackContentConverter\ContentPatcher\Patchers\PatcherAbstract` class, which holds most of the patchers' standard and reusable functionality.

This is the last "dirty" (not prettified, not CodeSniffer formatted) commit to be merged to the `master`. It was written quickly, to create and present the patching foundation. But the next commit would definitely apply formatting throughout this codebase.

A note: it is very probable that the `ImgPatcher:: patch_blocks_contents()` will be further simplified, and its code reused more (what's best to be done here would become more obvious when the other patchers get implemented). But the reason for this kind of structure is that it gives every single patcher completely independent patching logic, so that they may be very different in how they apply the patch. If we should run into extreme cases, each patch could do something completely different to the source to complete its patching, not following a strict formula or set of steps -- this flexibility is the biggest plus of this structure, I'd say.

### CODE REVIEW

File: lib/content-patcher/patchers/class-patcher-abstract.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	/**
+	 * Runs preg_replace.
+	 *
+	 * @param string $pattern_match
+	 * @param string $replacement
+	 * @param string $subject
+	 *
+	 * @return string|string[]|null preg_replace's default return.
+	 */
+	protected function apply_patch_block_element( $pattern_match, $replacement, $subject ) {
```
Comment (jeffersonrabb) [inline lib/content-patcher/patchers/class-patcher-abstract.php]: This looks like a pure reimplementation of `preg_replace` - do you foresee implementations of this method that contain more than the one line?

**Comment by iuravic** [comment]
> I like the approach. Left a bunch of fairly minor comments. The biggest concern I have is how we'll verify the regexes and test for false positives and other oddities. If we end up with a large number of patchers we'll have a lot of regexes to keep track of. Any thoughts on how to robustly test them for accuracy?

Excellent question! Exactly what I was thinking to solve with tests I was about to add in the future. But here they are now: [the tests for this ImgPatcher class](https://github.com/Automattic/newspack-content-converter/pull/5/files#diff-a1fb0849829c7d5798b1d222bb8e7162R30).

And here's the results of writing these: I discovered a bug, and also found an edge case which wasn't covered (which we might not need at the moment, but may come in handy, as a security precaution).

So by writing simple unit tests like these, we can make sure all the base functionality is covered and correct, and they'll keep checking that it doesn't get broken in by future modifications.

File: assets/src/utilities/index.js [hunk 2]
```diff
@@ -83,33 +83,33 @@ export function getPostContentById( id ) {
...
  *
- * @param data
+ * @param String html HTML source before conversion.
  * @returns {Promise<any> | Promise}
  */
-export function insertClassicBlockWithContent( data ) {
+export function insertClassicBlockWithContent( html ) {
 	return new Promise( function( resolve, reject ) {
-		const html = data.content.rendered;
 		var block = createBlock( 'core/freeform' );
```
[the remark this answers is not in the document]
Comment (iuravic) [reply assets/src/utilities/index.js]: Applied.

File: lib/class-converter.php [hunk 3]
```diff
@@ -19,20 +21,23 @@ class Converter {
...
+	/**
+	 * @var PatchHandler
+	 */
+	private $patcher_handler;
+
 	/**
 	 * Initialize the converter.
 	 */
-	public function init() {
+	public function __construct( PatchHandler $patcher_handler ) {
```
[the remark this answers is not in the document]
Comment (iuravic) [reply lib/class-converter.php]: This is the classic Dependency Injection pattern in action. By "injecting" the object, rather than instantiating it within the class, we can easily mock it in tests, more easily replace/refactor it. Besides it keeps these two services more decoupled.

File: assets/src/utilities/index.js [hunk 4]
```diff
@@ -42,17 +42,17 @@ export function runSinglePost( postId ) {
 		.then( () => {
 			return getPostContentById( postId );
 		} )
-		.then( data => {
-			return insertClassicBlockWithContent( data );
+		.then( html => {
```
[the remark this answers is not in the document]
Comment (iuravic) [reply assets/src/utilities/index.js]: Thank you! Applied.

File: lib/class-converter.php [hunk 5]
```diff
@@ -92,17 +97,36 @@ function() {
 	 * @param WP_REST_Request $params Params: 'id' Post ID, 'content' Post content.
 	 */
 	public function update_converted_post_content( $params ) {
-		$id      = $params->get_param( 'id' );
-		$content = $params->get_param( 'content' );
+		$postId         = $params->get_param( 'postId' );
```
[the remark this answers is not in the document]
Comment (iuravic) [reply lib/class-converter.php]: Good idea! Switched the content param to snake_case.

File: assets/src/utilities/index.js [hunk 6]
```diff
@@ -158,7 +162,7 @@ export function updatePost( data, id ) {
 			Accept: 'application/json, text/javascript, */*; q=0.01',
 			'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
 		},
-		body: `id=${ id }&content=${ dataEncoded }`,
+		body: `postId=${ postId }&content_blocks=${ blocksEncoded }&content_html=${ htmlEncoded }`,
```
[the remark this answers is not in the document]
Comment (iuravic) [reply assets/src/utilities/index.js]: Switched to the object instead of string.

File: lib/content-patcher/class-patch-handler.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	/**
+	 * @see \NewspackContentConverter\ContentPatcher\PatchHandlerInterface::run_all_patches
+	 *
+	 * @param string $html_content
+	 * @param string $block_content
+	 *
+	 * @return string|null
+	 */
+    public function run_all_patches( string $html_content, string $block_content ) : ?string {
+	    $patched_block_content = $block_content;
```
[the remark this answers is not in the document]
Comment (iuravic) [reply lib/content-patcher/class-patch-handler.php]: No special reason. Removed it.

File: lib/content-patcher/class-patch-handler.php [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * @package NewspackContentConverter\ContentPatcher
+ */
+class PatchHandler implements PatchHandlerInterface {
+
+	/**
+	 * @var array List of active content patchers.
+	 */
+	private $classes = [
+		'\NewspackContentConverter\ContentPatcher\Patchers\ImgPatcher',
```
[the remark this answers is not in the document]
Comment (iuravic) [reply lib/content-patcher/class-patch-handler.php]: My best guess at the moment (having focused on getting the conversion flow to work, and on the plugin concept much more than studying the content data), from 25 to 50.

File: lib/content-patcher/interface-patch-handler.php [hunk 9]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace NewspackContentConverter\ContentPatcher;
+
+interface PatchHandlerInterface {
```
[the remark this answers is not in the document]
Comment (iuravic) [reply lib/content-patcher/interface-patch-handler.php]: I am thinking about this one, and believe that we could remove it. I'll gladly remove it, but would rather wait a bit: to first write very simple tests for the PatchHandler, and the Plugin class which uses it, as well as to write some more patchers -- so that we don't remove it now, and need to add it later. But at this moment, I believe we'll be able to discard it.

File: lib/content-patcher/patchers/class-patcher-abstract.php [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+	/**
+	 * Validates the results of match searches.
+	 *
+	 * @param array $matches_html   Regex matches in HTML source.
+	 * @param array $matches_blocks Regex matches in blocks content.
+	 *
+	 * @return bool Do the matches correspond.
+	 */
+	protected function validate_html_and_block_matches( $matches_html, $matches_blocks ) {
```
[the remark this answers is not in the document]
Comment (iuravic) [reply lib/content-patcher/patchers/class-patcher-abstract.php]: Several cases come to my mind. First is that a different number is found. Second is that an element could be converted differently under different contexts (eg. when standalone HTML element, or nested within a certain structure then converted differently because of that structure, or simply because of a bug in conversion). These are wide shots, but these kind of checks make room for custom logic now. We might remove them, but I'd like to see a few patchers written before doing more refactoring.

File: lib/content-patcher/patchers/class-patcher-abstract.php (hunk 1 above)
Comment (iuravic) [reply lib/content-patcher/patchers/class-patcher-abstract.php]: Yes, exactly. I think we'll be able to optimize and refactor once we have several more patchers in place, hopefully of different types and scopes, and they'll show us how to improve this. My idea was to keep it simple, point the way, and keep it easily changeable for the next patchers. But flexible and neat, before all.

File: lib/content-patcher/patchers/class-imgpatcher.php [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+	protected $patterns = [
+		'match_html_element'     => '/<img.*?>/xim',
+		'match_blocks_element'   => '/<img.*?>/im',
+		'match_attribute_value'  => '/
+			.*?             # anything at the beginning of the string
+			height          # then locate the height attribute
+			\s*             # followed by zero or more spaces
+			=               # the equals char
+			\s*             # once again, with possible zero or more spaces
+			["\\\'+]?       # the attribute value starts with a double or a single quote
```
[the remark this answers is not in the document]
Comment (iuravic) [reply lib/content-patcher/patchers/class-imgpatcher.php]: Thank you for spotting this! Yes, both the `+` and the `\` are extra there. They didn't get a chance to do any damage (and probably wouldn't in the future), but they're wrong.

I cleaned them up now. And I also simplified the attribute value matching/capturing part, from `([^"\\']+.*?)` to `([^"']+)`, since only a `+` should do (meaning one or more char except the quote).

**Comment by iuravic** [comment]
> ...regex is not really the correct tool for parsing/fixing HTML, as HTML is not a regular language...

Point taken, and I agree with you. It would be much more readable to use a library such as a DOMDocument. However it seems to me that it's not actually possible in our case. Gutenberg persists attributes in comments, and we're actually going to be tweaking code that isn't really DOM. Gutenberg blocks' content uses it's own markup in some places, and because of this, the DOMDocument approach would be useful at best for half of the conversion and the other half would need to be a regex.
